planner/core/expression_rewriter.go

```diff
@@ -796,11 +796,22 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
 		er.err = err
 		return v, true
 	}
+
+	// If the leftKey and the rightKey have different charsets or collations, don't convert the sub-query to an inner-join
+	// since when converting we will add a distinct-agg upon the right child and this distinct-agg doesn't have the right collation.
+	// To keep it simple, we forbid this converting if they have different collations.
+	collFlag := true
+	lt, rt := lexpr.GetType(), rexpr.GetType()
+	if lt.EvalType() == types.ETString && rt.EvalType() == types.ETString &&
```

Comment:
Please consider `CompatibleCollate`, I think this function is enough, we don't need to check the charset any more.

Reply:
fixed, PTAL!